File: libraries/react-native/ios/HybridNotifications.swift
import Foundation
import NitroModules
import UserNotifications
import UIKit

final class HybridNotifications: HybridNotificationsSpec {
    private var listeners: [String: (NativeNotificationEvent) -> Void] = [:]
    private lazy var collectorSubscription = VoidhashPushCollector.shared.subscribe { [weak self] event in
        guard let self else { return }
        let kind: NativeNotificationEventKind
        let errorCode: String?
        switch event {
        case .tokenChanged:
            kind = .tokenchanged
            errorCode = nil
        case .registrationError(let code):
            kind = .registrationerror
            errorCode = code
        }
        let observed = NativeNotificationEvent(
            id: "notification_\(UUID().uuidString.lowercased())",
            kind: kind,
            occurredAt: ISO8601DateFormatter().string(from: Date()),
            protectedPayloadRef: nil,
            pushNotificationSendId: nil,
            link: nil,
            errorCode: errorCode
        )
        self.listeners.values.forEach { $0(observed) }
    }

    init() { _ = collectorSubscription }

    deinit { VoidhashPushCollector.shared.unsubscribe(collectorSubscription) }

    func getPermissionStatus() throws -> Promise<String> {
        Promise.async {
            let settings = await UNUserNotificationCenter.current().notificationSettings()
            switch settings.authorizationStatus {
            case .authorized: return "authorized"
            case .provisional, .ephemeral: return "provisional"
            case .denied: return "denied"
            case .notDetermined: return "notDetermined"
            @unknown default: return "notDetermined"
            }
        }
    }

    func requestPermission(provisional: Bool) throws -> Promise<String> {
        Promise.async {
            var options: UNAuthorizationOptions = [.alert, .badge, .sound]
            if provisional { options.insert(.provisional) }
            let granted = try await UNUserNotificationCenter.current().requestAuthorization(options: options)
            if granted {
                await MainActor.run { UIApplication.shared.registerForRemoteNotifications() }
            }
            return granted ? (provisional ? "provisional" : "authorized") : "denied"
        }
    }

    func getToken() throws -> Promise<NativePushToken> {
        Promise.async {
            guard let observed = VoidhashPushCollector.shared.currentToken() else {
                throw RuntimeError.error(withMessage: "PUSH_TOKEN_NOT_OBSERVED")
            }
            return NativePushToken(
                token: observed.token,
                provider: .apns,
                environment: observed.environment == .development ? .development : .production
            )
        }
    }

    func setBadgeCount(count: Double) throws -> Promise<Void> {
        guard count >= 0, count.rounded() == count else {
            throw RuntimeError.error(withMessage: "INVALID_BADGE_COUNT")
        }
        return Promise.async {
            if #available(iOS 16.0, *) {
                try await UNUserNotificationCenter.current().setBadgeCount(Int(count))
            } else {
                await MainActor.run { UIApplication.shared.applicationIconBadgeNumber = Int(count) }
            }
        }
    }

    func subscribe(subscriptionId: String, listener: @escaping (NativeNotificationEvent) -> Void) throws {
        listeners[subscriptionId] = listener
    }

    func unsubscribe(subscriptionId: String) throws { listeners.removeValue(forKey: subscriptionId) }
}

File: libraries/react-native/android/src/main/java/com/margelo/nitro/voidhash/HybridNotifications.kt
package com.margelo.nitro.voidhash

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.util.concurrent.ConcurrentHashMap
import com.margelo.nitro.voidhash.measurement.VoidhashPushCollector

class HybridNotifications : HybridNotificationsSpec() {
    private val listeners = ConcurrentHashMap<String, (NativeNotificationEvent) -> Unit>()
    private val collectorSubscriptionId = "hybrid-${System.identityHashCode(this)}"

    init {
        VoidhashPushCollector.subscribe(collectorSubscriptionId) { event ->
            val kind = when (event.kind) {
                "received" -> NativeNotificationEventKind.RECEIVED
                "opened" -> NativeNotificationEventKind.OPENED
                "tokenChanged" -> NativeNotificationEventKind.TOKENCHANGED
                else -> NativeNotificationEventKind.REGISTRATIONERROR
            }
            val bridged = NativeNotificationEvent(
                event.id,
                kind,
                event.occurredAt,
                event.protectedPayloadRef,
                event.pushNotificationSendId,
                event.link,
                event.errorCode,
            )
            listeners.values.forEach { it(bridged) }
        }
    }

    override fun getPermissionStatus(): Promise<String> = Promise.async {
        val context = NitroModules.applicationContext
            ?: throw IllegalStateException("NITRO_CONTEXT_UNAVAILABLE")
        if (Build.VERSION.SDK_INT < 33) "notRequired"
        else if (context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) "authorized"
        else if (context.getSharedPreferences("voidhash-notifications", 0).getBoolean("permission-requested", false)) "denied"
        else "notDetermined"
    }

    override fun requestPermission(provisional: Boolean): Promise<String> = Promise.async {
        if (Build.VERSION.SDK_INT < 33) return@async "notRequired"
        val context = NitroModules.applicationContext
            ?: throw IllegalStateException("NITRO_CONTEXT_UNAVAILABLE")
        if (context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
            return@async "authorized"
        }
        val activity = context.currentActivity
            ?: throw IllegalStateException("PERMISSION_REQUEST_REQUIRES_ACTIVITY")
        context.getSharedPreferences("voidhash-notifications", 0)
            .edit().putBoolean("permission-requested", true).apply()
        activity.runOnUiThread {
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                0x5648,
            )
        }
        repeat(300) {
            if (context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
                return@async "authorized"
            }
            Thread.sleep(100)
        }
        "denied"
    }

    override fun getToken(): Promise<NativePushToken> = Promise.async {
        val observed = VoidhashPushCollector.currentToken()
            ?: throw IllegalStateException("PUSH_TOKEN_NOT_OBSERVED")
        NativePushToken(observed.token, NativePushProvider.FCM, NativePushEnvironment.PRODUCTION)
    }

    override fun setBadgeCount(count: Double): Promise<Unit> = Promise.async {
        require(count >= 0 && count % 1.0 == 0.0) { "INVALID_BADGE_COUNT" }
        throw IllegalStateException("BADGE_COUNT_UNSUPPORTED_BY_ANDROID")
    }

    override fun subscribe(subscriptionId: String, listener: (NativeNotificationEvent) -> Unit) {
        listeners[subscriptionId] = listener
    }

    override fun unsubscribe(subscriptionId: String) {
        listeners.remove(subscriptionId)
    }
}
